File: cpp/subsets.cc
/*
 * Filename: subsets.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Sat 01 Aug 2015 08:53:54 PM EDT
 */

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
	public:
		vector<vector<int> > subsets(
				vector<int>& nums)
		{
			vector<vector<int> > res;

			if (!nums.size())
			{
				return res;
			}

			vector<int> sub;
			res.push_back(sub);
			sort(nums.begin(), nums.end());
			subsets_aux(nums, -1, res, sub);

			return res;
		}

	private:
		void subsets_aux(
				vector<int>& nums,
				int i,
				vector<vector<int> >& res,
				vector<int>& sub)
		{
			for (vector<int>::size_type j = i+1; j < nums.size(); j++)
			{
				sub.push_back(nums[j]);
				res.push_back(sub);
				subsets_aux(nums, j, res, sub);
				sub.pop_back();
			}
		}
};

File: java/Subsets.java
/*
 * Filename: Combinations.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Thu 21 May 2015 07:50:03 PM EDT
 */

public class Solution {
	public List<List<Integer>> subsets(int[] nums) {
		List<List<Integer>> res = new ArrayList<List<Integer>>();

		if (nums == null || nums.length == 0) {
			return res;
		}

		Arrays.sort(nums);
		for (int i = 0; i <= nums.length; i++) {
			res.addAll(combine(nums, nums.length, i));
		}

		res.add(new ArrayList<Integer>());
		return res;
	}

	private List<List<Integer>> combine(int[] nums, int n, int k) {
		List<List<Integer>> res = new ArrayList<List<Integer>>();

		if (n <= 0 || k <= 0) {
			return res;
		}

		if (k == 1) {
			for (int i = 0; i < n; i++) {
				List<Integer> comb = new ArrayList<Integer>();
				comb.add(nums[i]);
				res.add(comb);
			}

			return res;
		}

		if (k == n) {
			List<Integer> comb = new ArrayList<Integer>();
			for (int i = 0; i < n; i++) {
				comb.add(nums[i]);
			}
			res.add(comb);

			return res;
		}

		List<List<Integer>> list1 = combine(nums, n-1, k);
		List<List<Integer>> list2 = combine(nums, n-1, k-1);

		res.addAll(list1);
		for (List<Integer> comb: list2) {
			comb.add(nums[n-1]);
			res.add(comb);
		}

		return res;
	}
}
